### PR TITLE
OCEANWATER-565_Convert_GSAP_CSV_Input_to_Actual_Power_Input_Parameters

### PR DESCRIPTION
**Pull request by jfugate1**
## Linked Issues:
| EPIC ⚡| [Oceanwater-84 / Power System Modeling](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-84) |
| :---------- | :---------- |
| Jira Ticket 🎟️   | [OCEANWATER-565 / Convert GSAP CSV Input to Actual Power Input Parameters ](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-565) |
| Jira Ticket 🎟️   |[OCEANWATER-629 / Address or hide warning messages produced in power_system_node](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-629)
| Github :octocat:  | # |

## Summary of Changes
* Replaced precalculated csv dependency with rostopic subscription to average mechanical power
* Placeholding approximations for voltage and temperature inputs (randomly generated for now)
* Used data to feed into GSAP prognostics algorithm
* Update README.md

## Test
1. Complie, run OW
```bash
roslaunch ow (atacma_y1a.launch | europa_terminator_workspace.launch)
```
2. Subscribe to state_of_charge, remaining_useful_life, and battery_temperature and verify their output
```bash
rostopic echo /power_system_node/state_of_charge
rostopic echo /power_system_node/remaining_useful_life
rostopic echo /power_system_node/battery_temperature
```

### CODE REVIEW

**Review by Samahu — DISMISSED** [review]
It seems that you missed to include the file: `gsap_input_data.csv` as a result I am getting a plenty of "Unable to open data file" messages in the logs when I try to run the simulation.

**Comment by jfugate1** [comment]
> It seems that you missed to include the file: `gsap_input_data.csv` as a result I am getting a plenty of "Unable to open data file" messages in the logs when I try to run the simulation.

Interesting... The "gsap_input_data.csv" file is something that is generated by the code itself. It shouldn't have to be there initially. I will look into this.

**Comment by jfugate1** [comment]
> It seems that you missed to include the file: `gsap_input_data.csv` as a result I am getting a plenty of "Unable to open data file" messages in the logs when I try to run the simulation.

Ahh I see... when I removed all the content from the "data" folder, the folder was removed as well, which is causing the issue. I will fix this.

**Comment by jfugate1** [comment]
@Samahu issue should be resolved with my latest commit

File: ow_power_system/src/power_system_node.cpp [hunk 1]
```diff
@@ -13,92 +13,116 @@
...
+
+    // Set mechanical power value to rostopic subscription
+    float mech_power = msg->data; // [W]
+
+    // Create temperature estimate with pseudorandom noise generator
+    float min_temp = 17.5; // minimum temp = 17.5 deg. C
+    float max_temp = 21.5; // maximum temp = 21.5 deg. C
+
+    // Temperature estimate based on pseudorandom noise and fixed range
+    float temp_est = min_temp + static_cast <float> (rand()) / (static_cast <float> (RAND_MAX/(max_temp - min_temp)));
```
Comment (Samahu) [inline ow_power_system/src/power_system_node.cpp:56]: I don't know where you got this formula to perform range random but it wouldn't give you the desired results
```suggestion
    float temp_est = min_temp + static_cast <float> (rand()) / RAND_MAX * (max_temp - min_temp);
```

File: ow_power_system/src/power_system_node.cpp [hunk 2]
```diff
@@ -13,92 +13,116 @@
...
+    rul_msg.data = 0;  // immediately set to a good default
+    std_msgs::Float64 soc_msg;
+    soc_msg.data = 0.0;
+    std_msgs::Float64 tempbat_msg;
+    tempbat_msg.data = 0.0;
+
+    int temp_index = 1; // Set to 1 for now (constant vector). This will change to median SOC or RUL index or fixed percentile
+
+    // Read battery data from newly created file.
+    auto power_data = read_file(csv_path);
```
Comment (Samahu) [inline ow_power_system/src/power_system_node.cpp:109]: I know exactly what you doing here and although it makes good re-use of existing code this is considered lazy programming :facepalm:

**Review by Samahu — CHANGES_REQUESTED** [review]
You don't need to write the data to a file and reload it just to re-use the **read_file** method.
Simply examine what **read_file** function is doing and fill the same data structure right in memory .. avoiding the write/read to disk which is super slow compared to copying data in memory. This is critically important since this is performed within a callback.

**Comment by lanseafood** [comment]
@jfugate1  regarding "Subscribe to power_system_node/state_of_charge, remaining_useful_life, and battery_temperature and verify output "

are there specific values we should be verifying?

File: ow_power_system/src/power_system_node.cpp (hunk 1 above)
Comment (jfugate1) [reply ow_power_system/src/power_system_node.cpp]: Resolved in latest commit

**Comment by jfugate1** [comment]
> @jfugate1 regarding "Subscribe to power_system_node/state_of_charge, remaining_useful_life, and battery_temperature and verify output "
> 
> are there specific values we should be verifying?

Good question @lanseafood ! There aren't necessarily specific values to look for since the output now depends on the power output of the system. The values should make sense based on the status of the lander. When actions requiring mechanical power are performed, the RUL and SOC should progressively decrease, whereas they will decrease at a much slower pace when no actions are being performed.

File: ow_power_system/src/power_system_node.cpp [hunk 3]
```diff
@@ -14,91 +14,131 @@
...
-
-  if (!csv_path_param_exist)
+using namespace chrono;
+
+double update_count = 0.00; // Timestep initialization
+double battery_lifetime = 2738.0; // Estimate of battery lifetime (seconds)
+Datum<double> init_power = 0.00; // Power initialization
+Datum<double> init_temp = 20.00; // Temperature initialization
+Datum<double> init_voltage = 4.10; // Voltage initialization
+auto init_time = system_clock::now(); // Timestep initialization
```
Comment (Samahu) [inline ow_power_system/src/power_system_node.cpp:26]: Convert these to member variables of **SubscribeAndPublish** class and initialize them in the constructor

File: ow_power_system/src/power_system_node.cpp [hunk 4]
```diff
@@ -14,91 +14,131 @@
...
+
+class SubscribeAndPublish
```
Comment (Samahu) [inline ow_power_system/src/power_system_node.cpp:28]: Use a more descriptive name of the class .. for example **PowerSystemNode**..

File: ow_power_system/src/power_system_node.cpp [hunk 5]
```diff
@@ -14,91 +14,131 @@
...
-
-  if (update_rate_param_exist)
+  }
+  void powerCallback(const std_msgs::Float64::ConstPtr& msg)
   {
-    ROS_WARN_NAMED("power_system_node", "Overriding the default power_update_rate!");
-    // Validated the parameter
-    if (power_update_rate == 0)
+    using namespace std::chrono;
+    ros::NodeHandle nh;
```
Comment (Samahu) [inline ow_power_system/src/power_system_node.cpp:47]: Unreferenced variable
```suggestion
```

File: ow_power_system/src/power_system_node.cpp [hunk 6]
```diff
@@ -14,91 +14,131 @@
...
-  bool update_rate_param_exist = nh.param("power_update_rate", power_update_rate_override, 0.1);
-
-  if (update_rate_param_exist)
+  }
+  void powerCallback(const std_msgs::Float64::ConstPtr& msg)
   {
-    ROS_WARN_NAMED("power_system_node", "Overriding the default power_update_rate!");
-    // Validated the parameter
-    if (power_update_rate == 0)
+    using namespace std::chrono;
```
Comment (Samahu) [inline ow_power_system/src/power_system_node.cpp:46]: duplicate statement .. you have already defined this clause early in file
```suggestion
```

File: ow_power_system/src/power_system_node.cpp [hunk 7]
```diff
@@ -14,91 +14,131 @@
...
+    Datum<double> temperature = temp_est;
+    temperature.setTime(timestamp);
+    Datum<double> voltage = voltage_est;
+    voltage.setTime(timestamp);
+
+    // Populate initialization data structure
+    initialization_data.insert({MessageId::Watts, init_power});
+    initialization_data.insert({MessageId::Centigrade, init_temp});
+    initialization_data.insert({MessageId::Volts, init_voltage});
+    power_data.push_back(initialization_data);
```
Comment (Samahu) [inline ow_power_system/src/power_system_node.cpp:112]: Why do you think that it is necessary to constantly submit the initialization data? I don't understand it .. can you clarify?

Comment (Samahu) [reply ow_power_system/src/power_system_node.cpp:112]: Just to note that this wasn't the case when we were reading from the csv file .. So I am very skeptic about what is happening in this part of the code.

File: ow_power_system/src/power_system_node.cpp [hunk 8]
```diff
@@ -14,91 +14,131 @@
...
+    init_voltage = voltage_est;
+    init_time = timestamp;
+
+    // Create a configuration from a file
+    string config_path = ros::package::getPath("ow_power_system")+"/config/example.cfg";
+    ConfigMap config(config_path);
+    // Contruct a new prognoser using the prognoser factory. The prognoser
+    // will automatically construct an appropriate model, observer and predictor
+    // based on the values specified in the config.
+    auto prognoser = PrognoserFactory::instance().Create("ModelBasedPrognoser", config);
```
Comment (Samahu) [inline ow_power_system/src/power_system_node.cpp:133]: Why do you need to create and destroy a `prognoser` on each callback?!

**Review by Samahu — CHANGES_REQUESTED** [review]
- [ ] rename **SubscribeAndPublish** to **PowerSystemNode**
- [ ] Apply [OceanWATERS Code Style](https://babelfish.arc.nasa.gov/confluence/display/OCEANWATERS/Coding+Style+and+Practices)
- [ ] declare prognoser (named m_prognoser) as a member variable for the PowerSystemNode
- [ ] Initialize m_prognoser in the constructor (or anywhere before you subscribe to the mechanical power topic)
- [ ] Re-use m_prognoser within the callback
- [ ] Revise the code with these two points in mind
  - [ ] Is it necessary to submit the initialization data each time?
  - [ ] If it is not necessary then consider dropping the **for** loop

File: ow_power_system/src/power_system_node.cpp [hunk 9]
```diff
@@ -119,34 +160,45 @@ int main(int argc, char* argv[]) {
...
     }
-        
   }
+
+private:
+  ros::NodeHandle nh;
+  ros::Publisher SOC_pub;
+  ros::Publisher RUL_pub;
+  ros::Publisher TempBat_pub;
+  ros::Subscriber MechPower_sub;
```
[the remark this answers is not in the document]
Comment (jfugate1) [reply ow_power_system/src/power_system_node.cpp]: Resolved in latest commit.

File: ow_power_system/src/power_system_node.cpp (hunk 3 above)
Comment (jfugate1) [reply ow_power_system/src/power_system_node.cpp]: Resolved in latest commit.

File: ow_power_system/src/power_system_node.cpp (hunk 4 above)
Comment (jfugate1) [reply ow_power_system/src/power_system_node.cpp]: Resolved in latest commit.

File: ow_power_system/src/power_system_node.cpp (hunk 5 above)
Comment (jfugate1) [reply ow_power_system/src/power_system_node.cpp]: Resolved in latest commit.

File: ow_power_system/src/power_system_node.cpp (hunk 6 above)
Comment (jfugate1) [reply ow_power_system/src/power_system_node.cpp]: Resolved in latest commit.

File: ow_power_system/src/power_system_node.cpp (hunk 7 above)
Comment (jfugate1) [reply ow_power_system/src/power_system_node.cpp]: Addressed in latest commit.

File: ow_power_system/src/power_system_node.cpp (hunk 8 above)
Comment (jfugate1) [reply ow_power_system/src/power_system_node.cpp]: Addressed in latest commit.

**Comment by jfugate1** [comment]
Please note that the latest commit resolves several of the issues described by @Samahu in the previous commit from a stylistic standpoint. However, the functionality needs work. The code compiles, but a crash occurs within the power system node when running due to the way in which the prognoser is initialized.

**Comment by lanseafood** [comment]
> Please note that the latest commit resolves several of the issues described by @Samahu in the previous commit from a stylistic standpoint. However, the functionality needs work. The code compiles, but a crash occurs within the power system node when running due to the way in which the prognoser is initialized.

Is that crash something you wanted to resolve in this PR or another?

**Comment by lanseafood** [comment]
Ran into these, seems like nothing is being published?

mlma@hoshi:~/oceanwaters_ws$ rostopic echo /power_system_node/state_of_charge
WARNING: no messages received and simulated time is active.
Is /clock being published?

mlma@hoshi:~/oceanwaters_ws$ rostopic echo /power_system_node/battery_temperature 
WARNING: no messages received and simulated time is active.
Is /clock being published?

mlma@hoshi:~/oceanwaters_ws$ rostopic echo /power_system_node/remaining_useful_life 
WARNING: no messages received and simulated time is active.
Is /clock being published?

**Comment by jfugate1** [comment]
> Ran into these, seems like nothing is being published?
> 
> mlma@hoshi:~/oceanwaters_ws$ rostopic echo /power_system_node/state_of_charge
> WARNING: no messages received and simulated time is active.
> Is /clock being published?
> 
> mlma@hoshi:~/oceanwaters_ws$ rostopic echo /power_system_node/battery_temperature
> WARNING: no messages received and simulated time is active.
> Is /clock being published?
> 
> mlma@hoshi:~/oceanwaters_ws$ rostopic echo /power_system_node/remaining_useful_life
> WARNING: no messages received and simulated time is active.
> Is /clock being published?

@lanseafood yes, Ussama was going to add a commit to fix these issues. I was just documenting the current status so he knows. Sorry, should've given you a heads up!

**Comment by lanseafood** [comment]
> @lanseafood yes, Ussama was going to add a commit to fix these issues. I was just documenting the current status so he knows. Sorry, should've given you a heads up!

No worries!

**Comment by Samahu** [comment]
> > Please note that the latest commit resolves several of the issues described by @Samahu in the previous commit from a stylistic standpoint. However, the functionality needs work. The code compiles, but a crash occurs within the power system node when running due to the way in which the prognoser is initialized.
> 
> Is that crash something you wanted to resolve in this PR or another?

I'll try to address them as part of this PR .. it maybe till Thursday until I submit a new request for review.

**Comment by Samahu** [comment]
To reviewers: I have completed the described work for this ticket and the power_system_node should run live. I did refactor parts of the code while doing so.
Additionally, I have added the 3 following tickets to the current sprint:
- [OCEANWATER-651 / Read initial battery voltage and battery temperature from config](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-651)
- [OCEANWATER-652 / Remove the strong dependency of PowerSystemNode on the mechanical power topic](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-652)
- [OCEANWATER-653 / Add the ability to apply additional load on the battery at runtime](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-653)

@jfugate1 I can not add you as a reviewer of this PR since you created the pull request but you should be able to give feedback as any other reviewer. I have also added Mike

**Comment by jfugate1** [comment]
> To reviewers: I have completed the described work for this ticket and the power_system_node should run live. I did refactor parts of the code while doing so.
> Additionally, I have added the 3 following tickets to the current sprint:
> 
> * [OCEANWATER-651 / Read initial battery voltage and battery temperature from config](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-651)
> * [OCEANWATER-652 / Remove the strong dependency of PowerSystemNode on the mechanical power topic](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-652)
> * [OCEANWATER-653 / Add the ability to apply additional load on the battery at runtime](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-653)
> 
> @jfugate1 I can not add you as a reviewer of this PR since you created the pull request but you should be able to give feedback as any other reviewer. I have also added Mike

Thanks, Ussama! I will get this review done as soon as possible and comment accordingly. I also figured it would make sense for you to eliminate the stylistic errors that produce warnings when building the power system node (if you haven't already done so).

**Comment by Samahu** [comment]
> > To reviewers: I have completed the described work for this ticket and the power_system_node should run live. I did refactor parts of the code while doing so.
> > Additionally, I have added the 3 following tickets to the current sprint:
> > 
> > * [OCEANWATER-651 / Read initial battery voltage and battery temperature from config](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-651)
> > * [OCEANWATER-652 / Remove the strong dependency of PowerSystemNode on the mechanical power topic](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-652)
> > * [OCEANWATER-653 / Add the ability to apply additional load on the battery at runtime](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-653)
> > 
> > @jfugate1 I can not add you as a reviewer of this PR since you created the pull request but you should be able to give feedback as any other reviewer. I have also added Mike
> 
> Thanks, Ussama! I will get this review done as soon as possible and comment accordingly. I also figured it would make sense for you to eliminate the stylistic errors that produce warnings when building the power system node (if you haven't already done so).

I haven't yet but I'll get to that

**Comment by jfugate1** [comment]
Verified the power system node now works as advertised. Once we can get the style casting warnings removed, I believe this will be ready to merge.

**Comment by Samahu** [comment]
> Verified the power system node now works as advertised. Once we can get the style casting warnings removed, I believe this will be ready to merge.

I have resolved [OCEANWATER-629 / Address or hide warning messages produced in power_system_node](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-629).
All of these warnings except one are outside our code. I addressed the one that was in our code and for the rest I disable the cmake flag causing this issue.

**Comment by jfugate1** [comment]
> > Verified the power system node now works as advertised. Once we can get the style casting warnings removed, I believe this will be ready to merge.
> 
> I have resolved [OCEANWATER-629 / Address or hide warning messages produced in power_system_node](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-629).
> All of these warnings except one are outside our code. I addressed the one that was in our code and for the rest I disable the cmake flag causing this issue.

Nice! Confirmed. Consider the PR approved on my end.

File: ow_power_system/include/power_system_node.h [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+#include <ros/ros.h>
+#include <chrono>
+
+#include "PrognoserFactory.h"
+
+class PowerSystemNode
+{
+public:
+  PowerSystemNode();
```
Comment (kmdalal) [inline ow_power_system/include/power_system_node.h:12]: It's good to also be explicit about the copy constructor, assignment operator, and destructor.  I.e. define them or 'delete' them or a comment that you want to use the defaults.

File: ow_power_system/include/power_system_node.h [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+  void Run();
+
+private:
+  void powerCallback(const std_msgs::Float64::ConstPtr& msg);
+
+private:
```
Comment (kmdalal) [inline ow_power_system/include/power_system_node.h:19]: A comment I've made before, but it's odd to see repeated "private:" etc.  Not idiomatic C++.

File: ow_power_system/include/power_system_node.h (hunk 10 above)
Comment (Samahu) [reply ow_power_system/include/power_system_node.h:12]: Yes, originally there was some initialization in the constructor. But I have moved things around and the constructor is no longer utilized .. I will omit it!

Comment (Samahu) [reply ow_power_system/include/power_system_node.h]: addressed

File: ow_power_system/include/power_system_node.h (hunk 11 above)
Comment (Samahu) [reply ow_power_system/include/power_system_node.h]: Removed the extra private

**Review by kmdalal — APPROVED** [review]
Made some minor code comments.  The power node worked as described, and the published values looked reasonable.  Great to see this in the system!

**Pull request merged**